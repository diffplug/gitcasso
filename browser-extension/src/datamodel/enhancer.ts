import type { OverType } from '../overtype/mock-overtype'

/**
 * stores enough info about the location of a draft to:
 * - display it in a table
 * - reopen the draft in-context
 */
export interface CommentSpot {
  unique_key: string
  type: string
}

/** wraps the textareas of a given platform with Gitcasso's enhancements */
export interface CommentEnhancer<T extends CommentSpot = CommentSpot> {
  /** guarantees to only return a type within this list */
  forCommentTypes(): string[]
  /**
   * whenever a new `textarea` is added to any webpage, this method is called.
   * if we return non-null, then we become the handler for that text area.
   */
  tryToEnhance(textarea: HTMLTextAreaElement): [OverType, T] | null

  generateIcon(context: T): string
  generateDisplayTitle(context: T): string
  buildUrl(context: T, withDraft?: boolean): string
}
