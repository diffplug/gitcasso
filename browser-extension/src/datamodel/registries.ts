import type { OverType } from '../overtype/mock-overtype'
import type { CommentEnhancer, CommentSpot } from './enhancer'
import { GitHubHandler as GitHubEnhancer } from './handlers/github-handler'
import { RedditHandler as RedditEnhancer } from './handlers/reddit-handler'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  element: HTMLTextAreaElement
  context: T
  handler: CommentEnhancer<T>
  overtype: OverType
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer<any>>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubEnhancer())
    this.register(new RedditEnhancer())
  }

  private register<T extends CommentSpot>(handler: CommentEnhancer<T>): void {
    this.enhancers.add(handler)
  }

  getHandlerForType(type: string): CommentEnhancer<any> | null {
    for (const handler of this.enhancers) {
      if (handler.forCommentTypes().includes(type)) {
        return handler
      }
    }
    return null
  }

  identifyTextarea(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | null {
    for (const handler of this.enhancers) {
      try {
        const result = handler.tryToEnhance(textarea)
        if (result) {
          const [overtype, context] = result
          return { context, element: textarea, handler, overtype }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  getAllHandlers(): CommentEnhancer<any>[] {
    return Array.from(this.enhancers)
  }

  getCommentTypesForHandler(handler: CommentEnhancer<any>): string[] {
    return handler.forCommentTypes()
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea<any>>()

  register<T extends CommentSpot>(textareaInfo: EnhancedTextarea<T>): void {
    this.textareas.set(textareaInfo.element, textareaInfo)
    // TODO: register as a draft in progress with the global list
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    if (this.textareas.has(textarea)) {
      // TODO: register as abandoned or maybe submitted with the global list
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | undefined {
    return this.textareas.get(textarea)
  }
}
