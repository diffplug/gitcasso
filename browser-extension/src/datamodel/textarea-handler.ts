
export interface CommentContext {
  unique_key: string;
  type: string;
}

export interface TextareaInfo<T extends CommentContext = CommentContext> {
  element: HTMLTextAreaElement;
  context: T;
  handler: TextareaHandler<T>;
}

export interface TextareaHandler<T extends CommentContext = CommentContext> {
  // Handler metadata
  forCommentTypes(): string[];
  // whenever a new `textarea` is added to any webpage, this method is called to try to find a handler for it
  identifyContextOf(textarea: HTMLTextAreaElement): T | null;
  
  // Popup functionality helpers
  generateDisplayTitle(context: T): string;
  generateIcon(context: T): string;
  buildUrl(context: T, withDraft?: boolean): string;
}

