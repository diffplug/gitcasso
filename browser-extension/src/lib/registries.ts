import type { OverType } from '../overtype/mock-overtype'
import type { CommentEnhancer, CommentSpot } from './enhancer'
import { GitHubAddCommentEnhancer } from './enhancers/github'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  textarea: HTMLTextAreaElement
  spot: T
  handler: CommentEnhancer<T>
  overtype: OverType
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer<any>>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubAddCommentEnhancer())
  }

  private register<T extends CommentSpot>(handler: CommentEnhancer<T>): void {
    this.enhancers.add(handler)
  }

  tryToEnhance(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | null {
    for (const handler of this.enhancers) {
      try {
        const result = handler.tryToEnhance(textarea)
        if (result) {
          const [overtype, spot] = result
          return { handler, overtype, spot, textarea }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  getEnhancerCount(): number {
    return this.enhancers.size
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea<any>>()

  register<T extends CommentSpot>(textareaInfo: EnhancedTextarea<T>): void {
    this.textareas.set(textareaInfo.textarea, textareaInfo)
    // TODO: register as a draft in progress with the global list
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    if (this.textareas.has(textarea)) {
      // TODO: register as abandoned or maybe submitted with the global list
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | undefined {
    return this.textareas.get(textarea)
  }
}
