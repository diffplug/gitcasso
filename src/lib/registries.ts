import type { OverTypeInstance } from 'overtype'
import OverType from 'overtype'
import type {
  CommentEnhancer,
  CommentEvent,
  CommentEventType,
  CommentSpot,
  StrippedLocation,
} from './enhancer'
import { CommentEnhancerMissing } from './enhancers/CommentEnhancerMissing'
import { GitHubEditEnhancer } from './enhancers/github/GitHubEditEnhancer'
import { GitHubIssueAppendEnhancer } from './enhancers/github/GitHubIssueAppendEnhancer'
import { GitHubIssueCreateEnhancer } from './enhancers/github/GitHubIssueCreateEnhancer'
import { GitHubPrAppendEnhancer } from './enhancers/github/GitHubPrAppendEnhancer'
import { GitHubPrCreateEnhancer } from './enhancers/github/GitHubPrCreateEnhancer'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  textarea: HTMLTextAreaElement
  spot: T
  enhancer: CommentEnhancer<T>
  overtype: OverTypeInstance
  cleanup?: () => void
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer>()
  byType = new Map<string, CommentEnhancer>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubEditEnhancer())
    this.register(new GitHubIssueAppendEnhancer())
    this.register(new GitHubIssueCreateEnhancer())
    this.register(new GitHubPrAppendEnhancer())
    this.register(new GitHubPrCreateEnhancer())
    const textColor = 'rgb(31, 35, 40)'
    const headingColor = 'rgb(174, 52, 151)'
    OverType.setTheme({
      colors: {
        blockquote: 'rgb(89, 99, 110)',
        code: '#59636e',
        codeBg: '#f6f8fa',
        cursor: '#000000',
        em: 'rgb(126, 123, 255)',
        h1: headingColor,
        h2: headingColor,
        h3: headingColor,
        hr: '#5a7a9b',
        link: 'rgb(9, 105, 218)',
        selection: 'rgba(0, 123, 255, 0.3)',
        strong: 'rgb(45, 1, 142)',
        syntaxMarker: textColor,
        text: textColor,
      },
      name: 'custom-github',
    })
  }

  private register<T extends CommentSpot>(enhancer: CommentEnhancer<T>): void {
    this.enhancers.add(enhancer)
    for (const spotType of enhancer.forSpotTypes()) {
      this.byType.set(spotType, enhancer)
    }
  }

  enhancerFor<T extends CommentSpot>(spot: T): CommentEnhancer<T> {
    return (this.byType.get(spot.type) || new CommentEnhancerMissing()) as CommentEnhancer<T>
  }

  tryToEnhance(textarea: HTMLTextAreaElement, location: StrippedLocation): EnhancedTextarea | null {
    for (const enhancer of this.enhancers) {
      try {
        const spot = enhancer.tryToEnhance(textarea, location)
        if (spot) {
          const { instance: overtype, cleanup } = enhancer.enhance(textarea, spot)
          this.handleDelayedValueInjection(overtype)
          return { enhancer, overtype, spot, textarea, ...(cleanup && { cleanup }) }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  private handleDelayedValueInjection(overtype: OverTypeInstance): void {
    // GitHub sometimes injects textarea content after a delay
    // We need to trigger OverType to update its preview after such injections
    // https://github.com/diffplug/gitcasso/issues/46
    setTimeout(() => {
      overtype.updatePreview()
    }, 100)
    setTimeout(() => {
      overtype.updatePreview()
    }, 200)
    setTimeout(() => {
      overtype.updatePreview()
    }, 400)
    setTimeout(() => {
      overtype.updatePreview()
    }, 800)
  }

  getEnhancerCount(): number {
    return this.enhancers.size
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea>()
  private eventSender: (event: CommentEvent) => void = () => {}

  setCommentEventSender(sendEvent: (event: CommentEvent) => void): void {
    this.eventSender = sendEvent
  }

  private sendEvent(eventType: CommentEventType, enhanced: EnhancedTextarea): void {
    this.eventSender({
      draft: enhanced.textarea.value,
      spot: enhanced.spot,
      type: eventType,
    })
  }

  register<T extends CommentSpot>(enhanced: EnhancedTextarea<T>): void {
    this.textareas.set(enhanced.textarea, enhanced)
    enhanced.textarea.addEventListener('blur', () => {
      this.sendEvent('LOST_FOCUS', enhanced)
    })
    this.sendEvent('ENHANCED', enhanced)
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    const enhanced = this.textareas.get(textarea)
    if (enhanced) {
      if (enhanced.cleanup) {
        enhanced.cleanup()
      }
      this.sendEvent('DESTROYED', enhanced)
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea | undefined {
    return this.textareas.get(textarea)
  }

  tabLostFocus(): void {
    for (const enhanced of this.textareas.values()) {
      this.sendEvent('LOST_FOCUS', enhanced)
    }
  }
}
