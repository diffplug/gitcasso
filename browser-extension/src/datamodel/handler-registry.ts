import { CommentType, CommentContext, TextareaHandler, TextareaInfo } from './textarea-handler';
import { GitHubHandler } from '../handlers/github-handler';
import { RedditHandler } from '../handlers/reddit-handler';

export class HandlerRegistry {
  private handlers = new Set<TextareaHandler<any>>();

  constructor() {
    // Register all available handlers
    this.register(new GitHubHandler());
    this.register(new RedditHandler());
  }

  private register<T extends CommentContext>(handler: TextareaHandler<T>): void {
    this.handlers.add(handler);
  }

  getHandlerForType(type: CommentType): TextareaHandler<any> | null {
    for (const handler of this.handlers) {
      if (handler.forCommentTypes().includes(type)) {
        return handler;
      }
    }
    return null;
  }

  identifyAll(): TextareaInfo<any>[] {
    const allTextareas: TextareaInfo<any>[] = [];
    
    for (const handler of this.handlers) {
      try {
        const textareas = handler.identify();
        allTextareas.push(...textareas);
      } catch (error) {
        console.warn('Handler failed to identify textareas:', error);
      }
    }
    
    return allTextareas;
  }

  getAllHandlers(): TextareaHandler<any>[] {
    return Array.from(this.handlers);
  }

  getCommentTypesForHandler(handler: TextareaHandler<any>): CommentType[] {
    return handler.forCommentTypes();
  }
}