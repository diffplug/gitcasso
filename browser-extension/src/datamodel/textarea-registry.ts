import { TextareaInfo, CommentContext } from './textarea-handler';

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, TextareaInfo<any>>();

  register<T extends CommentContext>(textareaInfo: TextareaInfo<T>): void {
    this.textareas.set(textareaInfo.element, textareaInfo);
  }

  get(textarea: HTMLTextAreaElement): TextareaInfo<any> | undefined {
    return this.textareas.get(textarea);
  }
}