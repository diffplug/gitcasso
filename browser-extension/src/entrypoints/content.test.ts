import { beforeEach, describe, expect, it, vi } from 'vitest'
import { EnhancerRegistry, TextareaRegistry } from '../datamodel/registries'

// Mock WXT's defineContentScript global
vi.stubGlobal('defineContentScript', vi.fn())

describe('GitHub PR Content Script', () => {
  let enhancers: EnhancerRegistry
  let enhancedTextareas: TextareaRegistry
  let mockTextarea: HTMLTextAreaElement

  beforeEach(() => {
    // Reset DOM and registries for each test
    document.body.innerHTML = ''
    enhancers = new EnhancerRegistry()
    enhancedTextareas = new TextareaRegistry()

    // Mock window.location for GitHub PR page
    Object.defineProperty(window, 'location', {
      writable: true,
      value: {
        hostname: 'github.com',
        pathname: '/diffplug/selfie/pull/517',
        href: 'https://github.com/diffplug/selfie/pull/517'
      }
    })

    // Create a mock textarea element that mimics GitHub's PR comment box
    mockTextarea = document.createElement('textarea')
    mockTextarea.name = 'comment[body]'
    mockTextarea.placeholder = 'Leave a comment'
    mockTextarea.className = 'form-control markdown-body'
    
    // Add it to a typical GitHub comment form structure
    const commentForm = document.createElement('div')
    commentForm.className = 'js-new-comment-form'
    commentForm.appendChild(mockTextarea)
    document.body.appendChild(commentForm)
  })

  it('should identify GitHub PR textarea and register it in TextareaRegistry', () => {
    // Simulate the content script's enhanceMaybe function
    const enhancedTextarea = enhancers.tryToEnhance(mockTextarea)
    
    expect(enhancedTextarea).toBeTruthy()
    expect(enhancedTextarea?.element).toBe(mockTextarea)
    expect(enhancedTextarea?.spot.type).toBe('GH_PR_ADD_COMMENT')
    
    // Register the enhanced textarea
    if (enhancedTextarea) {
      enhancedTextareas.register(enhancedTextarea)
    }
    
    // Verify it's in the registry
    const registeredTextarea = enhancedTextareas.get(mockTextarea)
    expect(registeredTextarea).toBeTruthy()
    expect(registeredTextarea?.element).toBe(mockTextarea)
  })

  it('should create correct GitHubContext spot for PR comment', () => {
    const enhancedTextarea = enhancers.tryToEnhance(mockTextarea)
    
    expect(enhancedTextarea).toBeTruthy()
    
    // Snapshot test on the spot value
    expect(enhancedTextarea?.spot).toMatchSnapshot('github-pr-517-spot')
    
    // Also verify specific expected values
    expect(enhancedTextarea?.spot).toMatchObject({
      type: 'GH_PR_ADD_COMMENT',
      domain: 'github.com',
      slug: 'diffplug/selfie',
      number: 517,
      unique_key: 'github:diffplug/selfie:pull:517',
      commentId: undefined
    })
  })

  it('should handle multiple textareas on the same page', () => {
    // Create a second textarea for inline code comments
    const codeCommentTextarea = document.createElement('textarea')
    codeCommentTextarea.className = 'form-control js-suggester-field'
    
    const inlineForm = document.createElement('div')
    inlineForm.className = 'js-inline-comment-form'
    inlineForm.appendChild(codeCommentTextarea)
    document.body.appendChild(inlineForm)

    // Test both textareas
    const mainCommentEnhanced = enhancers.tryToEnhance(mockTextarea)
    const codeCommentEnhanced = enhancers.tryToEnhance(codeCommentTextarea)

    expect(mainCommentEnhanced?.spot.type).toBe('GH_PR_ADD_COMMENT')
    expect(codeCommentEnhanced?.spot.type).toBe('GH_PR_CODE_COMMENT')
    
    // Register both
    if (mainCommentEnhanced) enhancedTextareas.register(mainCommentEnhanced)
    if (codeCommentEnhanced) enhancedTextareas.register(codeCommentEnhanced)

    // Verify both are registered
    expect(enhancedTextareas.get(mockTextarea)).toBeTruthy()
    expect(enhancedTextareas.get(codeCommentTextarea)).toBeTruthy()
  })

  it('should not enhance textarea on non-GitHub pages', () => {
    // Change location to non-GitHub site
    Object.defineProperty(window, 'location', {
      writable: true,
      value: {
        hostname: 'example.com',
        pathname: '/some/page',
        href: 'https://example.com/some/page'
      }
    })

    const enhancedTextarea = enhancers.tryToEnhance(mockTextarea)
    expect(enhancedTextarea).toBeNull()
  })
})