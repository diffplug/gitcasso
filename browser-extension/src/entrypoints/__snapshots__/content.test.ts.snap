// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitHub PR Content Script > should create correct GitHubContext spot for PR comment > github-pr-517-spot 1`] = `
{
  "commentId": undefined,
  "domain": "github.com",
  "number": 517,
  "slug": "diffplug/selfie",
  "type": "GH_PR_ADD_COMMENT",
  "unique_key": "github:diffplug/selfie:pull:517",
}
`;
